docs/resources/server/recipes/ingredients.md

```diff
@@ -1,143 +1,184 @@
 # Ingredients
 
-`Ingredient`s are predicate handlers for item-based inputs which check whether a certain `ItemStack` meets the condition to be a valid input in a recipe. All [vanilla recipes][recipes] that take inputs use an `Ingredient` or a list of `Ingredient`s, which is then merged into a single `Ingredient`.
+`Ingredient`s are used in [recipes] to check whether a given [`ItemStack`][itemstack] is a valid input for the recipe. For this purpose, `Ingredient` implements `Predicate<ItemStack>`, and `#test` can be called to confirm if a given `ItemStack` matches the ingredient.
 
-## Custom Ingredients
+The `Ingredient` class is final, which has several drawbacks for us. Most notably, this makes it impossible to extend `Ingredient` for custom ingredient types. NeoForge works around this by introducing the `ICustomIngredient` interface. This is not a direct replacement for regular `Ingredient`s, but we can convert to and from `Ingredient`s using `ICustomIngredient#toVanilla` and `Ingredient#getCustomIngredient`, respectively.
```

Comment:
`Ingredient` being final is not necessarily a downside for us. In the past, the `final` was ATed away and custom ingredients extended `Ingredient` directly. This however was a complete mess in several places, which is why custom ingredients were rewritten to use the mostly detached `ICustomIngredient` interface. This paragraph should be rewritten to not hammer on "Ingredient being final sucks".